documentation/dataplane/dash-flow/DASH_FLOW_SAI.md

```diff
@@ -0,0 +1,615 @@
+
+
+
+## Flow state definition
+
+We define the flow state as a hash table as follows:
+
+
+
+### SAI protobuf defination
+
+```c
+typedef struct _sai_protobuf_t
+{
+    /* String in protobuf format */
+    sai_s8_list_t protobuf;
+  
+} sai_protobuf_t;
+```
+
+
+
+### Flow table attribute definiation
+
+```c
+typedef enum _sai_flow_table_attr_t {
+    /* Start of attributes */
+    SAI_FLOW_TABLE_ATTR_START,
+
+    /* ENI for the flow table */
+    SAI_FLOW_TABLE_ATTR_ENI,
+
+    /* Size attribute of the flow table */
+    SAI_FLOW_TABLE_ATTR_SIZE,
+
+    /* Expiry time attribute for entries in the flow table */
+    SAI_FLOW_TABLE_ATTR_EXPIRE_TIME,
+
+    /* Behavior attribute to track TCP states */
+    SAI_FLOW_TABLE_ATTR_BEHAVIOR_TRACK_TCP_STATES,
+  
+    /* Reset a TCP connection if it is illegal */
+    SAI_FLOW_TABLE_ATTR_TCP_RESET,
+
+    /* Version of the flow table */
+    SAI_FLOW_TABLE_ATTR_VERSION,
+
+    /* Flag to specify how to use the key for the flow table */
+    SAI_FLOW_TABLE_ATTR_KEY_FLAG,
+
+    /* End of attributes */
+    SAI_FLOW_TABLE_ATTR_END
+     
+} _sai_flow_table_attr_t;
+
+```
+
+### Flow state definiation
+
+
+
+# Flow attribute
+
+```c
+typedef struct _sai_flow_state_t {
+    /* Flow state Key */
+    sai_dash_flow_key_t key;
+  
+    /* Flow state attribute list */
+    sai_attribute_t *attr_list;
+  
+    /* Flow state atrribute count */
+    uint32_t attr_count;
+
+} sai_dash_flow_state_t;
+
+```
+
+### Flow state key
+
+```c
+/**
+ * @brief Flow key for network and transport layer information.
+ */
+typedef struct _sai_dash_flow_key_t {
+    /* @brief Source IP address */
+    sai_ip_address_t src_ip;
+
+    /* @brief Destination IP address */
+    sai_ip_address_t dst_ip;
+
+    /* @brief IP Protocol (e.g., TCP(6) or UDP(17)) */
+    sai_uint8_t ip_protocol;
+
+    /* @brief Transport Layer Information (TCP/UDP/ICMP) */
+    sai_dash_ha_flow_l4_info_t l4_info;
+  
+    /* @brief VNI */
+  	sai_uint32_t vni;
+      
+} sai_dash_flow_key_t;
+
+/**
+ * @brief Union representing L4 flow information for various protocols.
+ */
+typedef union _sai_dash_ha_flow_l4_info_t {
+    /* @brief TCP/UDP information */
+    sai_dash_ha_flow_tcp_udp_info_t tcp_udp;
+
+    /* @brief ICMP information */
+    sai_dash_ha_flow_icmp_info_t icmp;
+} sai_dash_ha_flow_l4_info_t;
+
+/**
+ * @brief Structure representing L4 information for TCP and UDP flows.
+ */
+typedef struct _sai_dash_ha_flow_tcp_udp_info_t {
+    /* @brief Source port */
+    sai_uint16_t src_port;
+
+    /* @brief Destination port */
+    sai_uint16_t dst_port;
+} sai_dash_ha_flow_tcp_udp_info_t;
+
+/**
+ * @brief Structure representing L4 information for ICMP flows.
+ */
+typedef struct _sai_dash_ha_flow_icmp_info_t {
+    /* @brief ICMP Type */
+    sai_uint32_t type;
+
+    /* @brief ICMP code */
+    sai_uint32_t code;
+
+    /* @brief ICMP ID */
+    sai_uint32_t id;
+} sai_dash_ha_flow_icmp_info_t;
```

Comment:
I believe there are many things are not implemented in the current flow API PR, so we need to work on these.